crates/astria-sequencer-relayer/src/relayer/mod.rs

```diff
@@ -0,0 +1,304 @@
+
+        let validator = match (
+            &cfg.relay_only_validator_key_blocks,
+            &cfg.validator_key_file,
+        ) {
+            (true, Some(file)) => Some(
+                Validator::from_path(file).wrap_err("failed to get validator info from file")?,
+            ),
+            (true, None) => {
+                bail!("validator key file must be set if `disable_relay_all` is set")
+            }
+            (false, _) => None, // could also say that the file was unnecessarily set, but it's ok
+        };
+
+        let celestia_client::celestia_rpc::Client::Http(celestia) =
+            celestia_client::celestia_rpc::Client::new(
+                &cfg.celestia_endpoint,
+                Some(&cfg.celestia_bearer_token),
+            )
+            .await
+            .wrap_err("failed constructing celestia http client")?
+        else {
+            bail!("expected to get a celestia HTTP client, but got a websocket client");
+        };
+
+        let state = Arc::new(State::new());
+
+        Ok(Self {
+            sequencer,
+            sequencer_poll_period: Duration::from_millis(cfg.block_time),
+            celestia,
+            validator,
+            state,
+        })
+    }
+
+    pub(crate) fn subscribe_to_state(&self) -> watch::Receiver<state::StateSnapshot> {
+        self.state.subscribe()
+    }
+
+    /// Runs the relayer worker.
+    ///
+    /// # Errors
+    ///
+    /// Returns errors if sequencer block fetch or celestia blob submission
+    /// failed catastrophically (after `u32::MAX` retries).
+    #[instrument(skip_all)]
+    pub(crate) async fn run(self) -> eyre::Result<()> {
+        let latest_height_stream =
+            make_latest_height_stream(self.sequencer.clone(), self.sequencer_poll_period);
+        pin!(latest_height_stream);
+
+        let (submitter_task, submitter) =
+            spawn_submitter(self.celestia.clone(), self.state.clone());
+
+        let mut block_stream = read::BlockStream::new(self.sequencer.clone(), self.state.clone());
+        block_stream.set_block_time(self.sequencer_poll_period);
+
+        let mut forward_once_free: Fuse<
+            BoxFuture<Result<(), tokio::sync::mpsc::error::SendError<SequencerBlock>>>,
+        > = Fuse::terminated();
+
+        self.state.set_ready();
+
+        let reason = loop {
+            select!(
+                biased;
+
+                res = &mut forward_once_free, if !forward_once_free.is_terminated() => {
+                    block_stream.resume();
+                    // XXX: exiting because submitter only returns an error after u32::MAX
+                    // retries, which is practically infinity.
+                    if res.is_err() {
+                        let err = res.wrap_err(
+                            "submitter exited unexpectly while trying to forward block"
+                        );
+                        break err;
+                    }
```

Comment:
the errors returned here can be the blob submit channel closed, or it's at capacity, right? are there any other cases?